### PR TITLE
Support query pass-through for Elasticsearch connector

### PR DESCRIPTION
**Pull request by kasiafi**
This PR introduces `raw_query` Polymorphic Table Function for full query pass-through to Elasticsearch connector.

```markdown
# Elasticsearch
* Add `raw_query` table function for full query pass-through to the connector.
```

## Documentation

Documentation issue is filed: https://github.com/trinodb/trino/issues/13007.
When documenting, also deprecate the legacy query pass-through mechanism: https://trino.io/docs/current/connector/elasticsearch.html?highlight=elasticsearch#pass-through-queries

### CODE REVIEW

File: plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java [hunk 1]
```diff
@@ -1860,6 +1860,12 @@ public void testMissingIndex()
         assertTableDoesNotExist("nonexistent_table");
     }
 
+    @Test
+    public void testRemoteQueryTableFunction()
+    {
+        assertQuerySucceeds("SELECT * FROM TABLE(\"" + getSession().getCatalog().orElseThrow() + "\".\"system\".\"remote_query\"(\"schema\" => 'tpch', \"index\" => 'nation', \"query\" => '{\"query\": {\"match\": {\"name\": \"ALGERIA\"}}}'))");
```
Comment (ebyhr) [inline plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1866]: It would be better to verify the result in my opinion and how about formatting like this? 
```java
        assertThat(
                "SELECT * FROM " +
                        "TABLE(\"" + getSession().getCatalog().orElseThrow() + "\".\"system\".\"remote_query\"(" +
                        " \"schema\" => 'tpch'," +
                        " \"index\" => 'nation'," +
                        " \"query\" => '{\"query\": {\"match\": {\"name\": \"ALGERIA\"}}}'" +
                        "))")
                .matches(...);
```

Also, I would recommend adding empt result & failure case.

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ptf/RemoteQuery.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import static io.trino.spi.ptf.DescriptorMapping.EMPTY_MAPPING;
+import static io.trino.spi.ptf.ReturnTypeSpecification.GenericTable.GENERIC_TABLE;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class RemoteQuery
+        implements Provider<ConnectorTableFunction>
+{
+    public static final String NAME = "remote_query";
```
Comment (martint) [inline plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ptf/RemoteQuery.java:53]: Call this `query`

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java [hunk 3]
```diff
@@ -563,8 +568,8 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
                     if (!newRegexes.containsKey(columnName) && pattern instanceof Slice) {
                         IndexMetadata metadata = client.getIndexMetadata(handle.getIndex());
                         if (metadata.getSchema()
-                                    .getFields().stream()
-                                    .anyMatch(field -> columnName.equals(field.getName()) && field.getType() instanceof PrimitiveType && "keyword".equals(((PrimitiveType) field.getType()).getName()))) {
+                                .getFields().stream()
+                                .anyMatch(field -> columnName.equals(field.getName()) && field.getType() instanceof PrimitiveType && "keyword".equals(((PrimitiveType) field.getType()).getName()))) {
```
Comment (martint) [inline plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java:572]: Unrelated formatting change?

File: plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java (hunk 1 above)
Comment (martint) [reply plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1866]: That's intentional. Pass-through queries in Elasticsearch return a single VARCHAR column as a result.

**Review by martint — APPROVED** [review]
We should deprecate the old-style passthrough queries in favor of this.

File: plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java (hunk 1 above)
Comment (kasiafi) [reply plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java]: @ebyhr I added more tests including no match and failure.

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ptf/Query.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            QueryHandle handle = new QueryHandle(tableHandle);
+
+            return TableFunctionAnalysis.builder()
+                    .returnedType(returnedType)
+                    .handle(handle)
+                    .build();
+        }
+    }
+
+    public static class QueryHandle
```
Comment (martint) [inline plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ptf/Query.java:127]: `QueryFunctionHandle` maybe?

File: plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java [hunk 5]
```diff
@@ -1860,6 +1861,69 @@ public void testMissingIndex()
...
+            try {
+                assertQuery(actual, expected);
+                return;
+            }
+            catch (AssertionError e) {
+                // ignore, try other options
+            }
+        }
+        throw new AssertionError("Neither of expected options matched the actual result");
+    }
```
Comment (martint) [inline plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1925]: I don't think this is such a good idea. It masks the underlying assertion, so it's hard to tell what was different and it what way. It also doesn't scale when there are more than a couple of alternatives.

Instead, modify the query to normalize the results in some way. For instance, parse the resulting json, extract the values and sort them to get a consistent result.

**Review by martint — COMMENTED** [review]
Also, don't forget to update the docs or file an issue to do it later so we don't lose track of it.

**Comment by kasiafi** [comment]
@martint I applied comments. Please take a look.

**Review by martint — APPROVED** [review]

File: plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java [hunk 6]
```diff
@@ -1860,6 +1861,53 @@ public void testMissingIndex()
...
+    @Test
+    public void testQueryTableFunction()
+    {
+        // select single record
+        assertQuery("SELECT json_query(result, 'lax $[0][0].hits.hits._source') " +
+                        "FROM TABLE(elasticsearch.system.query(" +
+                        "schema => 'tpch', " +
+                        "index => 'nation', " +
+                        "query => '{\"query\": {\"match\": {\"name\": \"ALGERIA\"}}}')) t(result)",
+                "VALUES '{\"nationkey\":0,\"name\":\"ALGERIA\",\"regionkey\":0,\"comment\":\" haggle. carefully final deposits detect slyly agai\"}'");
```
Comment (findepi) [inline plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1873]: due to the fact that the function seems to dump ES result, i would consider calling a `raw_query`
(especially so that we keep `query` name free for future)

Comment (martint) [reply plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1873]: That's a good point. In ES parlance, queries are "search queries" (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-your-data.html), and there's a Query DSL (https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html) to formulate them (what we support here). They also have two other languages:
* EQL, for querying log and event data (https://www.elastic.co/guide/en/elasticsearch/reference/current/eql.html)
* SQL (https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-overview.html)

**Comment by kasiafi** [comment]
@martint @findepi I renamed the function to `raw_query`.

I also added a test  to compare the results between the legacy pass-through mechanism, and the new table function.

The old query pass-through will be deprecated in the documentation when we document the table function. @martint is there any other way we want to deprecate it than a note in the docs?

**Review by martint — APPROVED** [review]
